File: revit-cad/pyrevit/ARX.extension/lib/arx_rulecore/pattern_builder.py
"""Turn "an example value + the part that varies" into a regex.

Faithful port of the Rule Check tool's buildPattern (and its toRuns /
runToPattern / runToLoose helpers). Producing byte-identical patterns to the
PDF/JS tool is what lets a rule set authored once run unchanged in PDF, Revit
and CAD.
"""

from .util import escape_regex

RUN_CLASS_LABEL = {"digit": "digit", "upper": "uppercase letter", "lower": "lowercase letter"}


def classify_ch(ch):
    if "0" <= ch <= "9":
        return "digit"
    if "A" <= ch <= "Z":
        return "upper"
    if "a" <= ch <= "z":
        return "lower"
    return "literal"


def to_runs(text):
    """Group consecutive characters of the same class into runs."""
    runs = []
    for ch in text:
        cls = classify_ch(ch)
        if runs and runs[-1]["cls"] == cls:
            runs[-1]["text"] += ch
        else:
            runs.append({"cls": cls, "text": ch})
    return runs


def run_to_pattern(run):
    """Exact-length class token: e.g. a 3-digit run -> \\d{3}."""
    if run["cls"] == "literal":
        return escape_regex(run["text"])
    cls = {"digit": "\\d", "upper": "[A-Z]", "lower": "[a-z]"}[run["cls"]]
    return cls if len(run["text"]) == 1 else "%s{%d}" % (cls, len(run["text"]))


def run_to_loose(run):
    """Any-length class token, used by the looser locator: e.g. \\d+."""
    if run["cls"] == "literal":
        return escape_regex(run["text"])
    return {"digit": "\\d+", "upper": "[A-Z]+", "lower": "[a-z]+"}[run["cls"]]


def run_to_desc(run):
    if run["cls"] == "literal":
        return 'the text "%s"' % run["text"]
    n = len(run["text"])
    return "%d %s%s" % (n, RUN_CLASS_LABEL[run["cls"]], "" if n == 1 else "s")


def build_pattern(example, variable_part, exact=False):
    """Return {valid, locate, explanation, warning, error} (port of buildPattern).

    - ``valid``  is anchored (^...$) and used to validate a captured field value.
    - ``locate`` is bounded by non-word lookarounds and is intentionally looser
      (same class skeleton, any lengths) so malformed instances are still found.
    """
    if not example:
        return {"error": "Enter an example value first."}

    if exact:
        valid = "^%s$" % escape_regex(example)
        locate = "(?<![A-Za-z0-9])%s(?![A-Za-z0-9])" % escape_regex(example)
        return {"valid": valid, "locate": locate,
                "explanation": 'Will match exactly: "%s"' % example,
                "warning": None, "error": None}

    if not variable_part:
        return {"error": 'Enter the part of the example that changes — '
                         'or set "exact" to true.'}

    index = example.find(variable_part)
    if index == -1:
        return {"error": '"%s" was not found inside the example value.' % variable_part}

    warning = None
    if example.find(variable_part, index + 1) != -1:
        warning = ('"%s" appears more than once in the example — the first '
                   "occurrence was used." % variable_part)

    prefix = example[:index]
    suffix = example[index + len(variable_part):]
    runs = to_runs(variable_part)

    valid = "^%s%s%s$" % (escape_regex(prefix),
                          "".join(run_to_pattern(r) for r in runs),
                          escape_regex(suffix))
    locate = "(?<![A-Za-z0-9])%s%s%s(?![A-Za-z0-9])" % (
        escape_regex(prefix),
        "".join(run_to_loose(r) for r in runs),
        escape_regex(suffix),
    )

    parts = []
    if prefix:
        parts.append('the text "%s"' % prefix)
    parts.extend(run_to_desc(r) for r in runs)
    if suffix:
        parts.append('the text "%s"' % suffix)

    return {"valid": valid, "locate": locate,
            "explanation": "Will match: " + " + ".join(parts),
            "warning": warning, "error": None}

File: revit-cad/dotnet/RuleCore/PatternBuilder.cs
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Arx.RuleCore
{
    public static class RegexUtil
    {
        private static readonly HashSet<char> Special =
            new HashSet<char>(".*+?^${}()|[]\\-");

        // Port of src/util.js escapeRegex — note spaces are NOT escaped.
        public static string EscapeRegex(string value)
        {
            var sb = new StringBuilder();
            foreach (var c in value ?? "")
            {
                if (Special.Contains(c)) sb.Append('\\');
                sb.Append(c);
            }
            return sb.ToString();
        }
    }

    // Port of the Rule Check tool's buildPattern (+ toRuns/runToPattern/runToLoose).
    public static class PatternBuilder
    {
        private sealed class Run { public string Cls; public string Text; }

        private static readonly Dictionary<string, string> ClassLabel =
            new Dictionary<string, string>
            {
                { "digit", "digit" }, { "upper", "uppercase letter" }, { "lower", "lowercase letter" }
            };

        private static string Classify(char ch)
        {
            if (ch >= '0' && ch <= '9') return "digit";
            if (ch >= 'A' && ch <= 'Z') return "upper";
            if (ch >= 'a' && ch <= 'z') return "lower";
            return "literal";
        }

        private static List<Run> ToRuns(string text)
        {
            var runs = new List<Run>();
            foreach (var ch in text)
            {
                var cls = Classify(ch);
                if (runs.Count > 0 && runs[runs.Count - 1].Cls == cls)
                    runs[runs.Count - 1].Text += ch;
                else
                    runs.Add(new Run { Cls = cls, Text = ch.ToString() });
            }
            return runs;
        }

        private static string RunToPattern(Run r)
        {
            if (r.Cls == "literal") return RegexUtil.EscapeRegex(r.Text);
            var cls = r.Cls == "digit" ? "\\d" : r.Cls == "upper" ? "[A-Z]" : "[a-z]";
            return r.Text.Length == 1 ? cls : cls + "{" + r.Text.Length + "}";
        }

        private static string RunToLoose(Run r)
        {
            if (r.Cls == "literal") return RegexUtil.EscapeRegex(r.Text);
            return r.Cls == "digit" ? "\\d+" : r.Cls == "upper" ? "[A-Z]+" : "[a-z]+";
        }

        private static string RunToDesc(Run r)
        {
            if (r.Cls == "literal") return "the text \"" + r.Text + "\"";
            int n = r.Text.Length;
            return n + " " + ClassLabel[r.Cls] + (n == 1 ? "" : "s");
        }

        public static PatternResult Build(string example, string variablePart, bool exact)
        {
            if (string.IsNullOrEmpty(example))
                return new PatternResult { Error = "Enter an example value first." };

            if (exact)
            {
                return new PatternResult
                {
                    Valid = "^" + RegexUtil.EscapeRegex(example) + "$",
                    Locate = "(?<![A-Za-z0-9])" + RegexUtil.EscapeRegex(example) + "(?![A-Za-z0-9])",
                    Explanation = "Will match exactly: \"" + example + "\"",
                };
            }

            if (string.IsNullOrEmpty(variablePart))
                return new PatternResult { Error = "Enter the part of the example that changes — or set exact to true." };

            int index = example.IndexOf(variablePart, System.StringComparison.Ordinal);
            if (index == -1)
                return new PatternResult { Error = "\"" + variablePart + "\" was not found inside the example value." };

            string warning = example.IndexOf(variablePart, index + 1, System.StringComparison.Ordinal) != -1
                ? "\"" + variablePart + "\" appears more than once in the example — the first occurrence was used."
                : null;

            string prefix = example.Substring(0, index);
            string suffix = example.Substring(index + variablePart.Length);
            var runs = ToRuns(variablePart);

            string valid = "^" + RegexUtil.EscapeRegex(prefix)
                         + string.Concat(runs.Select(RunToPattern))
                         + RegexUtil.EscapeRegex(suffix) + "$";
            string locate = "(?<![A-Za-z0-9])" + RegexUtil.EscapeRegex(prefix)
                          + string.Concat(runs.Select(RunToLoose))
                          + RegexUtil.EscapeRegex(suffix) + "(?![A-Za-z0-9])";

            var parts = new List<string>();
            if (prefix.Length > 0) parts.Add("the text \"" + prefix + "\"");
            parts.AddRange(runs.Select(RunToDesc));
            if (suffix.Length > 0) parts.Add("the text \"" + suffix + "\"");

            return new PatternResult
            {
                Valid = valid,
                Locate = locate,
                Explanation = "Will match: " + string.Join(" + ", parts),
                Warning = warning,
            };
        }
    }
}
